### PR TITLE
Add drawElementsInstanced session demo

### PR DESCRIPTION
**Pull request by vinsentli**
Add drawElementsInstanced session demo.

This demo will take effec after https://github.com/facebook/igl/pull/121 be merged.

metal : 
image

OpenGL:
image

OpengGLES:
223af76b7339beb07339c5ddb70ffa2d

Vulkan:
d223ce0a2dbfcb9e68fe157ed5e6ca49

### CODE REVIEW

File: shell/renderSessions/DrawInstancedSession.cpp [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+          struct VertexOut {
+            float4 position [[position]];
+            float3 uvw;
+          };
+
+          vertex VertexOut vertexShader(uint vid [[vertex_id]], ushort iid [[instance_id]],
+        constant VertexIn * vertices [[buffer(1)]], VertexIn in [[stage_in]]) {
+            VertexOut out;
+            //method one：
```
Comment (corporateshark) [inline shell/renderSessions/DrawInstancedSession.cpp:54]: Please remove unused code.

File: shell/renderSessions/DrawInstancedSession.cpp [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+    return igl::ShaderStagesCreator::fromModuleStringInput(device,
+                                                           getVulkanVertexShaderSource(),
+                                                           "main",
+                                                           "",
+                                                           getVulkanFragmentShaderSource(),
+                                                           "main",
+                                                           "",
+                                                           nullptr);
+    return nullptr;
+  // @fb-only
```
Comment (corporateshark) [inline shell/renderSessions/DrawInstancedSession.cpp:125]: Remove internal markup.

File: shell/renderSessions/DrawInstancedSession.cpp [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+          getVulkanVertexShaderSource(), std::regex("gl_VertexIndex"), "gl_VertexID");
+      auto codeVS2 =
+          std::regex_replace(codeVS1.c_str(), std::regex("460"), !usesOpenGLES ? "410" : "300 es");
+
+      auto codeFS = std::regex_replace(
+          getVulkanFragmentShaderSource(), std::regex("460"), !usesOpenGLES ? "410" : "300 es");
+
+      if (usesOpenGLES) {
+        codeVS2 =
+            std::regex_replace(codeVS2.c_str(), std::regex("layout \\(location=0\\) out"), "out");
```
Comment (corporateshark) [inline shell/renderSessions/DrawInstancedSession.cpp:148]: No need to use regexps for this. Use this instead (change to `const char*` where appropriate):
```
[[maybe_unused]] void stringReplaceAll(std::string& s,
                                       const std::string& searchString,
                                       const std::string& replaceString) {
  size_t pos = 0;
  while ((pos = s.find(searchString, pos)) != std::string::npos) {
    s.replace(pos, searchString.length(), replaceString);
  }
}
```
Please remove all `<regex>` usage.

File: shell/renderSessions/DrawInstancedSession.h [hunk 4]
```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+// @fb-only
```
Comment (corporateshark) [inline shell/renderSessions/DrawInstancedSession.h:8]: Remove internal markup in this new file.

**Review by corporateshark — CHANGES_REQUESTED** [review]

**Comment by vinsentli** [comment]
@corporateshark 
I have made the modifications according to your advice.

**Review by corporateshark — APPROVED** [review]
@vinsentli I will import this PR once https://github.com/facebook/igl/pull/121 is merged.

File: shell/renderSessions/DrawInstancedSession.cpp [hunk 5]
```diff
@@ -172,6 +172,16 @@ void DrawInstancedSession::initialize() noexcept {
   renderPass_.colorAttachments[0].storeAction = StoreAction::Store;
   renderPass_.colorAttachments[0].clearColor = getPlatform().getDevice().backendDebugColor();
   renderPass_.depthAttachment.loadAction = LoadAction::DontCare;
+
+  // Create Index Buffer
+  int16_t indexes[6] = {0, 1, 2, 3, 4, 5};
```
Comment (corporateshark) [inline shell/renderSessions/DrawInstancedSession.cpp:177]: `const`

File: shell/renderSessions/DrawInstancedSession.cpp [hunk 6]
```diff
@@ -251,7 +261,8 @@ void DrawInstancedSession::update(igl::SurfaceTextures surfaceTextures) noexcept
   commands->bindScissorRect(scissor);
   commands->pushDebugGroupLabel("Render Triangle", igl::Color(1, 0, 0));
   commands->bindVertexBuffer(1, *vertex_buffer_);
-  commands->draw(6, 100);
+  commands->bindIndexBuffer(*index_buffer_, IndexFormat::UInt16);
+  commands->drawIndexed(PrimitiveType::Triangle, 6, 100);
```
Comment (corporateshark) [inline shell/renderSessions/DrawInstancedSession.cpp:265]: This method is deprecated and will be removed. Use `drawIndexed(6, 100)`.

**Review by corporateshark — CHANGES_REQUESTED** [review]

**Comment by corporateshark** [comment]
@vinsentli please rebase onto the latest `main` branch.

**Comment by vinsentli** [comment]
> @vinsentli please rebase onto the latest `main` branch.

I rebase onto the latest `main` branch， noticed that there are quite a few exceptions.

So i create a new PR :  https://github.com/facebook/igl/pull/132

**Pull request closed without merging**